File: src/main/exporter.js
'use strict';

/**
 * CSV and Excel export.
 *
 * Both formats share one column layout so a spreadsheet built from either is
 * interchangeable.
 */

const fs = require('node:fs');
const ExcelJS = require('exceljs');

const clock = require('../shared/clock');

// Status stays the third column: the row tinting below finds it by position.
const HEADERS = [
  'Date',
  'Name',
  'Status',
  'In',
  'Break out',
  'Break in',
  'Out',
  'Hours worked',
  'Balance',
  'Reason for not coming',
  'Recorded at',
];

const HEADER_FILL = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF1F4E78' } };
const HEADER_FONT = { color: { argb: 'FFFFFFFF' }, bold: true };

const STATUS_FILL = {
  Present: 'FFE2F0D9',
  Late: 'FFFFF2CC',
  Absent: 'FFFBE5E5',
};

const toRecords = (rows) =>
  rows.map((r) => [
    r.date,
    r.name,
    r.status,
    clock.format(r.time_in),
    clock.format(r.break_out),
    clock.format(r.break_in),
    clock.format(r.time_out),
    clock.formatDuration(clock.workedMinutes(r)),
    clock.formatBalance(clock.balanceMinutes(r)),
    r.reason || '',
    r.recorded_at,
  ]);

function csvCell(value) {
  const text = String(value ?? '');
  return /[",\n\r]/.test(text) ? `"${text.replace(/"/g, '""')}"` : text;
}

function exportCsv(rows, destination) {
  const lines = [HEADERS.map(csvCell).join(',')];
  for (const record of toRecords(rows)) lines.push(record.map(csvCell).join(','));
  // The BOM makes Excel on Windows read accented names correctly.
  fs.writeFileSync(destination, '﻿' + lines.join('\r\n'), 'utf8');
  return destination;
}

function styleHeader(sheet) {
  const row = sheet.getRow(1);
  row.eachCell((cell) => {
    cell.fill = HEADER_FILL;
    cell.font = HEADER_FONT;
    cell.alignment = { horizontal: 'center', vertical: 'middle' };
  });
}

function autosize(sheet, min = 10, max = 45) {
  sheet.columns.forEach((column) => {
    let longest = 0;
    column.eachCell({ includeEmpty: false }, (cell) => {
      longest = Math.max(longest, String(cell.value ?? '').length);
    });
    column.width = Math.max(min, Math.min(max, longest + 3));
  });
}

async function exportXlsx(rows, destination, sheetTitle = 'Attendance') {
  const wb = new ExcelJS.Workbook();
  wb.creator = 'Attendance List';
  wb.created = new Date();

  const sheet = wb.addWorksheet(String(sheetTitle).slice(0, 31) || 'Attendance');
  sheet.addRow(HEADERS);
  for (const record of toRecords(rows)) sheet.addRow(record);

  styleHeader(sheet);

  // Tint each row by status so absences stand out when scanning.
  sheet.eachRow((row, index) => {
    if (index === 1) return;
    const colour = STATUS_FILL[row.getCell(3).value];
    if (!colour) return;
    row.eachCell((cell) => {
      cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: colour } };
    });
  });

  autosize(sheet);
  sheet.views = [{ state: 'frozen', ySplit: 1 }];
  sheet.autoFilter = { from: 'A1', to: { row: sheet.rowCount, column: HEADERS.length } };

  await wb.xlsx.writeFile(destination);
  return destination;
}

/** Workbook with the raw records plus a per-person summary and reasons. */
async function exportSummaryXlsx(rows, stats, reasons, destination) {
  const wb = new ExcelJS.Workbook();
  wb.creator = 'Attendance List';
  wb.created = new Date();

  const records = wb.addWorksheet('Records');
  records.addRow(HEADERS);
  for (const record of toRecords(rows)) records.addRow(record);
  styleHeader(records);
  autosize(records);
  records.views = [{ state: 'frozen', ySplit: 1 }];

  // Hours come from the records rather than the stats query: they are worked
  // out from four columns, which is arithmetic SQL should not be doing.
  const hoursByPerson = new Map();
  const shortByPerson = new Map();
  const overByPerson = new Map();
  for (const row of rows) {
    const hours = clock.workedHours(row);
    if (hours === null) continue;
    const add = (map, value) => map.set(row.person_id, (map.get(row.person_id) || 0) + value);
    add(hoursByPerson, hours);
    add(shortByPerson, clock.shortfallHours(row));
    add(overByPerson, clock.overtimeHours(row));
  }

  const round = (n) => Math.round((n || 0) * 100) / 100;

  const summary = wb.addWorksheet('Summary');
  summary.addRow([
    'Name', 'Recorded days', 'Present', 'Late', 'Absent', 'Attendance rate',
    'Hours worked', 'Short by', 'Overtime',
  ]);
  for (const s of stats) {
    const recorded = s.recorded || 0;
    const attended = (s.present || 0) + (s.late || 0);
    summary.addRow([
      s.name,
      recorded,
      s.present || 0,
      s.late || 0,
      s.absent || 0,
      recorded ? attended / recorded : 0,
      round(hoursByPerson.get(s.person_id)),
      round(shortByPerson.get(s.person_id)),
      round(overByPerson.get(s.person_id)),
    ]);
  }
  summary.getColumn(6).numFmt = '0%';
  for (const column of [7, 8, 9]) summary.getColumn(column).numFmt = '0.00';
  styleHeader(summary);
  autosize(summary);
  summary.views = [{ state: 'frozen', ySplit: 1 }];

  const reasonSheet = wb.addWorksheet('Reasons');
  reasonSheet.addRow(['Reason', 'Times given']);
  for (const r of reasons) reasonSheet.addRow([r.reason, r.n]);
  styleHeader(reasonSheet);
  autosize(reasonSheet);

  await wb.xlsx.writeFile(destination);
  return destination;
}

module.exports = { HEADERS, exportCsv, exportXlsx, exportSummaryXlsx };

File: app/exporter.py
"""CSV and Excel export.

Both formats share one column layout so a spreadsheet built from either is
interchangeable.
"""

from __future__ import annotations

import csv
import sqlite3
from pathlib import Path
from typing import Sequence

from openpyxl import Workbook
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.utils import get_column_letter

from . import db

HEADERS = ["Date", "Name", "Status", "Reason for not coming", "Recorded at"]

_HEADER_FILL = PatternFill("solid", fgColor="1F4E78")
_HEADER_FONT = Font(color="FFFFFF", bold=True)

_STATUS_FILL = {
    db.STATUS_PRESENT: PatternFill("solid", fgColor="E2F0D9"),
    db.STATUS_LATE: PatternFill("solid", fgColor="FFF2CC"),
    db.STATUS_ABSENT: PatternFill("solid", fgColor="FBE5E5"),
}


def _rows_to_records(rows: Sequence[sqlite3.Row]) -> list[list[str]]:
    return [
        [
            r["date"],
            r["name"],
            r["status"],
            r["reason"] or "",
            r["recorded_at"],
        ]
        for r in rows
    ]


def export_csv(rows: Sequence[sqlite3.Row], destination: Path) -> Path:
    destination = Path(destination)
    # utf-8-sig so Excel on Windows opens accented names correctly.
    with destination.open("w", newline="", encoding="utf-8-sig") as fh:
        writer = csv.writer(fh)
        writer.writerow(HEADERS)
        writer.writerows(_rows_to_records(rows))
    return destination


def export_xlsx(
    rows: Sequence[sqlite3.Row], destination: Path, sheet_title: str = "Attendance"
) -> Path:
    destination = Path(destination)
    wb = Workbook()
    ws = wb.active
    ws.title = sheet_title[:31] or "Attendance"

    ws.append(HEADERS)
    for cell in ws[1]:
        cell.fill = _HEADER_FILL
        cell.font = _HEADER_FONT
        cell.alignment = Alignment(horizontal="center", vertical="center")

    for record in _rows_to_records(rows):
        ws.append(record)

    # Tint each row by status so absences stand out when scanning.
    for row in ws.iter_rows(min_row=2, max_row=ws.max_row):
        fill = _STATUS_FILL.get(row[2].value)
        if fill:
            for cell in row:
                cell.fill = fill

    _autosize(ws)
    ws.freeze_panes = "A2"
    ws.auto_filter.ref = f"A1:{get_column_letter(len(HEADERS))}{ws.max_row}"

    wb.save(destination)
    return destination


def export_summary_xlsx(
    rows: Sequence[sqlite3.Row],
    stats: Sequence[sqlite3.Row],
    reasons: Sequence[sqlite3.Row],
    destination: Path,
) -> Path:
    """Workbook with the raw records plus a per-person summary and reasons."""
    destination = Path(destination)
    wb = Workbook()

    ws = wb.active
    ws.title = "Records"
    ws.append(HEADERS)
    for record in _rows_to_records(rows):
        ws.append(record)
    _style_header(ws)
    _autosize(ws)
    ws.freeze_panes = "A2"

    summary = wb.create_sheet("Summary")
    summary.append(["Name", "Recorded days", "Present", "Late", "Absent", "Attendance rate"])
    for s in stats:
        recorded = s["recorded"] or 0
        attended = (s["present"] or 0) + (s["late"] or 0)
        rate = (attended / recorded) if recorded else 0
        summary.append(
            [
                s["name"],
                recorded,
                s["present"] or 0,
                s["late"] or 0,
                s["absent"] or 0,
                rate,
            ]
        )
    for row in summary.iter_rows(min_row=2, max_row=summary.max_row, min_col=6, max_col=6):
        for cell in row:
            cell.number_format = "0%"
    _style_header(summary)
    _autosize(summary)
    summary.freeze_panes = "A2"

    reason_sheet = wb.create_sheet("Reasons")
    reason_sheet.append(["Reason", "Times given"])
    for r in reasons:
        reason_sheet.append([r["reason"], r["n"]])
    _style_header(reason_sheet)
    _autosize(reason_sheet)

    wb.save(destination)
    return destination


def _style_header(ws) -> None:
    for cell in ws[1]:
        cell.fill = _HEADER_FILL
        cell.font = _HEADER_FONT
        cell.alignment = Alignment(horizontal="center", vertical="center")


def _autosize(ws, minimum: int = 10, maximum: int = 45) -> None:
    for column in ws.columns:
        longest = max((len(str(c.value)) for c in column if c.value is not None), default=0)
        letter = get_column_letter(column[0].column)
        ws.column_dimensions[letter].width = max(minimum, min(maximum, longest + 3))
